pygeoapi/api.py

```diff
@@ -104,28 +113,320 @@
+    If `xml` was requested, we set the `Content-Type` ourselves. For the
+    standard formats, the `APIRequest` object sets the `Content-Type`.
+
+    .. code-block:: python
+
+       @pre_process
+       def example_method(self, request: Union[APIRequest, Any], custom_arg):
+          if not request.is_valid(['xml']):
+             return self.get_format_exception(request)
+
+          content_type = 'application/xml' if request.format == 'xml' else None
+          headers = request.get_response_headers(content_type)
+
+          # generate response_body here
+
+          return headers, 200, response_body
+
+    Note that you don't *have* to call :meth:`is_valid`, but that you can also
+    perform a custom check on the requested output format by looking at the
+    :attr:`format` property.
+    Other query parameters are available through the :attr:`params` property as
+    a `dict`. The request body is available through the :attr:`data` property.
+
+    .. note::   If the request data (body) is important, **always** create a
+                new `APIRequest` instance using the :meth:`with_data` factory
+                method.
+                The :func:`pre_process` decorator will use this automatically.
+
+    :param request:             The web platform specific Request instance.
+    :param supported_locales:   List or set of supported Locale instances.
+    """
+    def __init__(self, request, supported_locales):
+        # Set default request data
+        self._data = b''
+
+        # Copy request query parameters
+        self._args = self._get_params(request)
+
+        # Get path info
+        self._path_info = request.headers.environ['PATH_INFO'].strip('/')
+
+        # Extract locale from params or headers
+        self._raw_locale, self._locale = self._get_locale(request.headers,
+                                                          supported_locales)
+
+        # Determine format
+        self._format = self._get_format(request.headers)
+
+    @classmethod
+    def with_data(cls, request, supported_locales) -> 'APIRequest':
+        """ Factory class method to create an `APIRequest` instance with data.
+
+        If the request body is required, an `APIRequest` should always be
+        instantiated using this class method. The reason for this is, that the
+        Starlette request body needs to be awaited (async), which cannot be
+        achieved in the :meth:`__init__` method of the `APIRequest`.
+        However, `APIRequest` can still be initialized using :meth:`__init__`,
+        but then the :attr:`data` property value will always be empty.
+
+        :param request:             The web platform specific Request instance.
+        :param supported_locales:   List or set of supported Locale instances.
+        :returns:                   An `APIRequest` instance with data.
+        """
+        api_req = cls(request, supported_locales)
+        if hasattr(request, 'data'):
```

Comment:
@GeoSander how much is this strategy flexible and would that fit with other frameworks like Django (see #630)?

Reply:
Good point @francbartoli. The honest answer is: I don't know. However, looking at Django's [Request](https://docs.djangoproject.com/en/3.1/ref/request-response/) object, I see that all methods and/or properties are there to have the `APIRequest` object extract all the stuff we need.